DataDog/datadog-agent
# Persisting Third Party Integrations On Windows

### What does this PR do?

Persists Third party integrations on windows. 

### Motivation
https://datadoghq.atlassian.net/browse/WINA-1225
https://datadoghq.atlassian.net/browse/WINA-1252

Currently on upgrade and repair, third party integrations are uninstalled. This pull request will reinstall installed third party integrations on upgrade, and repairs. Also adds test for persistence feature on windows. 

This feature was implemented for other platforms in this #27859 

### Describe how you validated your changes


Manually tested with following steps:
 - Grab two MSI's from pipeline: `datadog-agent-` and  `datadog-agent-upgrade-test`
 - Install `datadog-agent-` MSI with flag `INSTALL_PYTHON_THIRD_PARTY_DEPS="1"`
   - EX: `msiexec /i datadog-agent-7.63.0-devel.git.170.2d9110f-1-x86_64.msi /l*v C:\base_log.txt  INSTALL_PYTHON_THIRD_PARTY_DEPS="1"`
 - Install `datadog-ping` integration:
   - `agent.exe integration install -t datadog-ping==1.0.2`
 - Upgrade to `datadog-agent-upgrade-test` MSI:
   - EX: `msiexec /i datadog-agent-upgrade-test-7.63.1-devel.git.170.2d9110f-1-x86_64.msi /l*v C:\upgrade_log.txt INSTALL_PYTHON_THIRD_PARTY_DEPS="1"`
 - Run `freeze` command and verify that `datadog-ping` is still installed
   - `.\agent.exe integration freeze`
 - Verify that integration is in the list
 
 
These steps are also covered by an added e2e test. 

### Possible Drawbacks / Trade-offs

Currently the pre-remove script runs on the following conditions: RemovingForUpgrade, uninstall, and maintenance. This is to support people upgrading by uninstalling -> installing. This is also how it works during fleet automation. So this will also persist integrations during fleet automation installs. 

Creates a protected folder to store everything in. If the protected folder is not owned by `SYSTEM` or `Administrator` the feature will not run. It also will not try to fix delete or fix that issue it will just fail. 

### Additional Notes


If the script fails, the install process will be affected. The result of the task is not ignored and the install fails on install. Should we fail the install if the script fails?

Rollback features added future PR in [rollback integrations windows](https://github.com/DataDog/datadog-agent/pull/33013)
